src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java

```diff
@@ -0,0 +1,276 @@
+        PrivateKey privateKey = loadPrivateKey(keyPath, keyPassword);
+
+        KeyStore keyStore = KeyStore.getInstance("JKS");
+        keyStore.load(null, null);
+        keyStore.setCertificateEntry("client-cert", clientCertificate);
+        keyStore.setKeyEntry("client-key", privateKey, keyPassword.toCharArray(),
+                new Certificate[] {clientCertificate});
+
+        KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        kmf.init(keyStore, keyPassword.toCharArray());
+        return kmf.getKeyManagers();
+    }
+
+    private static Certificate loadCertificate(String certificatePem) throws IOException, GeneralSecurityException {
+        CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");
+        InputStream certstream = fileToStream(certificatePem);
+        return certificateFactory.generateCertificate(certstream);
+    }
+
+    // PKCS#8 format
+    private static final String PEM_PRIVATE_START = "-----BEGIN PRIVATE KEY-----";
+    private static final String PEM_PRIVATE_END = "-----END PRIVATE KEY-----";
+    // PKCS#1 format
+    private static final String PEM_RSA_PRIVATE_START = "-----BEGIN RSA PRIVATE KEY-----";
+    // PVK format
+    private static final long PVK_MAGIC = 0xB0B5F11EL;
+    private static final byte[] RSA2_MAGIC = {82, 83, 65, 50};
+    private static final String RC4_ALG = "RC4";
+    private static final String RSA_ALG = "RSA";
+
+    private static PrivateKey loadPrivateKey(String privateKeyPemPath,
+            String privateKeyPassword) throws GeneralSecurityException, IOException, SQLServerException {
+        String privateKeyPem = getStringFromFile(privateKeyPemPath);
+
+        if (privateKeyPem.contains(PEM_PRIVATE_START)) { // PKCS#8 format
+            return loadPrivateKeyFromPKCS8(privateKeyPem);
+        } else if (privateKeyPem.contains(PEM_RSA_PRIVATE_START)) { // PKCS#1 format
+            return loadPrivateKeyFromPKCS1(privateKeyPem, privateKeyPassword);
+        } else {
+            return loadPrivateKeyFromPVK(privateKeyPemPath, privateKeyPassword);
+        }
+    }
+
+    private static PrivateKey loadPrivateKeyFromPKCS8(
+            String key) throws IOException, NoSuchAlgorithmException, InvalidKeySpecException {
+        key = key.replace(PEM_PRIVATE_START, "").replace(PEM_PRIVATE_END, "");
+        key = key.replaceAll("\\s", "");
+        byte[] pkcs8EncodedKey = Base64.getDecoder().decode(key);
+
+        KeyFactory factory = KeyFactory.getInstance(RSA_ALG);
+        return factory.generatePrivate(new PKCS8EncodedKeySpec(pkcs8EncodedKey));
+    }
+
+    private static PrivateKey loadPrivateKeyFromPKCS1(String key,
+            String keyPass) throws IOException, NoSuchAlgorithmException, InvalidKeySpecException {
+        try {
+            Security.addProvider(new BouncyCastleProvider());
+        } catch (SecurityException se) {
+            // fall through, provider already loaded

```

Comment:
Instead of relying on empty exception handling, is there a way to check if the provider is loaded, and use that ?

Reply:
Hi @saurabh500, in my opinion this is the best way to handle loading this dependency. The empty catch is there in case the user has loaded BouncyCastle themselves, and this is more efficient and reliable then trying to determine whether they've loaded BC or not.

Reply:
- SecurityException is thrown if a security manager exists and its method denies access to add a new provider. So there can be multiple reasons why this exception is thrown, not only when the provider exists.
- Why is this approach better than just checking if provider is loaded using `Security.getProvider(name)`?